File: associate-invoiced-products-to-account/functions/zoho-crm-api.js
const axios = require('axios');
const dotenv = require('dotenv');

dotenv.config();

class ZohoCrmApi {
    constructor() {
        this.access_token = this.refreshAccessTokens(process.env.CRM_REFRESH_TOKEN);
        this.conn = axios.create({
            baseURL: 'https://www.zohoapis.com/crm/v5/',
            headers: {
                'Authorization': `Zoho-oauthtoken ${this.access_token}`
            }
        });
    }

    // Refresh access token
    async refreshAccessTokens(refreshToken) {
        try {
            const response = await axios.post('https://accounts.zoho.com/oauth/v2/token', null, {
                params: {
                    refresh_token: refreshToken,
                    client_id: process.env.CLIENT_ID,
                    client_secret: process.env.CLIENT_SECRET,
                    grant_type: 'refresh_token'
                }
            });
            if (response.data.access_token) {
                return response.data.access_token;
            } else {
                throw new Error('Failed to get access token');
            }
        } catch (error) {
            console.error('Error refreshing access token:', error);
            throw error;
        }
    }

    // General API call function
    async apiCall(endpoint, method, config = {}) {
        try {
            const response = await this.conn({
                url: endpoint,
                method: method,
                ...config
            });
            return response.data;
        } catch (error) {
            console.error(`Error making API call to ${endpoint}:`, error);
            throw error;
        }
    }

    // Search function to search CRM modules
    async search(zModule, params) {
        const response = await this.apiCall(`${zModule}/search?criteria=${params}`, 'get');
        return response;
    }

    // CoQL query function
    async coql(selectQuery) {
        const response = await this.apiCall('coql', 'post', { select_query: selectQuery });
        return response;
    }

    // Post data in batches if payload is large
    async zohoCrmPost(url, type, payload) {
        const payloadLen = payload.length;
        const numOfRequests = Math.ceil(payloadLen / 100);
        let startIndex = 0;
        let endIndex = payloadLen > 100 ? 99 : payloadLen - 1;
        const respArray = [];

        for (let iteration = 1; iteration <= numOfRequests; iteration++) {
            const data = { data: payload.slice(startIndex, endIndex + 1) };
            const apiResp = await this.apiCall(url, type, data);
            respArray.push(apiResp.data);
            startIndex += 100;
            endIndex = iteration === numOfRequests ? payloadLen - 1 : endIndex + 100;
        }

        return respArray;
    }
}

module.exports = ZohoCrmApi;

File: Ruby/associate-invoiced-products-to-account/helpers/zoho-crm-api.rb
require_relative 'zoho'

class ZohoCrmApi < Zoho
  attr_accessor :conn

  def initialize
    access_token = self.refresh_access_tokens(ENV['CRM_REFRESH_TOKEN'])
    self.conn = Faraday.new 'https://www.zohoapis.com/crm/v5/' do |builder|
      builder.request :authorization , 'Zoho-oauthtoken', -> {access_token}
      builder.request :json
      builder.response :json
      builder.response :raise_error
      builder.response :logger
    end
    self
  end

  def search(z_module, params)
    response = self.api_call("#{z_module}/search?criteria=#{params}", 'get')
    response
  end

  def coql(select_query)
    self.api_call('coql', 'post', {select_query: select_query})
  end

  def zoho_crm_post(url, type, payload)
    payload_len = payload.length
    num_of_requests = (payload_len.to_f / 100 ).ceil
    start_index = 00
    end_index =
      if payload_len > 100
        99
      else
        payload_len - 1
      end
    resp_array = []

    for iteration in 1..num_of_requests do
      data = {data: payload[start_index..end_index]}
      api_resp = self.api_call(url, type, data)
      resp_array.append(api_resp['data'])
      start_index += 100
      end_index += 100
      next_iter = iteration + 1
      end_index = payload_len - 1 if next_iter == num_of_requests
    end
    resp_array
  end
end
